torch/_higher_order_ops/cond.py

```diff
@@ -429,30 +379,154 @@ def cond_fake_tensor_mode(mode, pred, true_fn, false_fn, operands):
+    Step 1: create unbacked symints for sizes that are different
+    along the same axis. For example:
+        a.size is [s0, 4, s0, 5, 4, 5]
+        b.size is [s1, 4, s2, 8, 4, 7]
+        merged_size will be [u0, 4, u1, u2, 4, u3], where
+        u0 has range [min(s0, s1), max(s0, s1)]
+        u1 has range [min(s0, s2), max(s0, s2)]
+        u2 has range [5, 8]
+        u3 has range [5, 7]
+    """
+    merged_size: list[Union[int, torch.SymInt]] = []
+    for s0, s1 in zip(a.size(), b.size()):
+        if SymIntEqByExpr(s0) == SymIntEqByExpr(s1):
+            merged_size.append(s0)
+        else:
+
+            def min_max(s0, s1):
+                def _bound(s: Union[int, torch.SymInt], lower_bound: bool):
+                    assert isinstance(s, (int, torch.SymInt))
+                    if isinstance(s, int):
+                        return s
+                    assert mode.shape_env is not None
+                    vrange = bound_sympy(s.node.expr, mode.shape_env.var_to_range)
+                    return vrange.lower if lower_bound else vrange.upper
+
+                return _bound(min(s0, s1), True), _bound(max(s0, s1), False)
+
+            assert mode.shape_env is not None
+            new_size = mode.shape_env.create_unbacked_symint()
+            mode.shape_env._update_var_to_range(
+                new_size.node.expr, ValueRanges(*min_max(s0, s1))
             )
-        true_meta = _extract_tensor_metadata(true_out)
-        false_meta = _extract_tensor_metadata(false_out)
-        if true_meta != false_meta:
-            raise torch._dynamo.exc.CondOpArgsMismatchError(
-                f"Expected each tensor to have same metadata but got:"
-                f"\n  {true_fn.__name__} returns {true_meta}"
-                f"\n  {false_fn.__name__} returns {false_meta}"
+            merged_size.append(new_size)
+
+    """
+    Step 2: Since tensor stride is a accumulative muliplication of the sizes (a permutated
+        ascending sequence), given a tensor size, stride and merged_size, we can re-map the
+        stride of the tensor using merged_size.
+
+        Without loss of generality, suppose we have a tenosr with:
+            size [3, 4, 3, 5, 4, 5],
+            stride (1200, 300, 1, 12, 3, 60),
+            merged_size [3, 4, 3, u0, 4, u0], where we bound 5 to u0.
+        The following code visit the strides in the ascending order: 1, 3, 12, 60, 300, 1200.
+        In each step, it checks whether the current stride is bounded or not and bound old stride
+        to new value. It also generates the next candidate by bounding stride[i] * size[i]
+        to bounded_stride * merged_size[i]. Concretely:
+        1. In 1st iteration, we bound 1 to 1, and set candidates[3] = 3
+        2. In 2nd iteration, we bound 3 to 3, and set candiates[12] = 12
+        3. In 3rd iteration, we bound 12 to 12, and set candidates[60] = 12 * u0
+        4. In 4th iteration, we bound 60 to 12 * u0, and set candidates[300] = 12 * u0 * u0
```

Comment:
We can only do this assuming the tensor is dense in memory, right? We should error if the tensor isn't dense in memory (is there a check for that somewhere?)

Reply:
I don't see a check for that